### DISCUSSION TITLE
Java.Lang.NoClassDefFoundError "Failed resolution of: Landroid/support/v4/util/SimpleArrayMap"

### DISCUSSION THREAD

**Discussion by losullivan59** [dotnet/maui#7652] [Q&A]
I'm brand new to dot net MAUI and Android development in general.

I'm trying to call an existing .aar library from a maui android application. The .aar came from an android studio java project (Kvaser canlib android 1.2.0, https://www.kvaser.com/download/). I have the source code for the library, and I was able to build it in Android Studio, to generate the aar. The android studio project consists of an app, as well as the library. The provided app seems to run well enough and calls the library without trouble.

I'd like to use the AAR from a MAUI app. I created an "Android Java Library Binding" for the AAR, and it seems to build ok. I can see everything I expect in the vs2022 object explorer. When I first go to use the library from a MAUI app, I get an exception of "Java.Lang.NoClassDefFoundError", "Failed resolution of: Landroid/support/v4/util/SimpleArrayMap".

From some Googling, my best guess is that it might be a problem with the android support library v4 (or lack thereof)? I tried to add Xamarin.Support.v4 nuget package, but got yet more exotic errors with low google search results (MSB4018, The "VerifyVersionsTask" task failed unexpectedly).

The AAR's android studio project was set up with the following sdk versions:
    compileSdkVersion = 23
    targetSdkVersion = 23
    buildToolsVersion = "28.0.3"
    minSdkVersionCanLib = 12
    minSdkVersionApp = 16

Also,  in the android studio project dependancies:
    api 'com.android.support:appcompat-v7:23.1.1'
    api 'com.android.support:support-annotations:23.1.1'

The android studio project, upon first opening form Kvaser's website, prompted me to upgrade gradle multiple times, until it seemed happy enough to build at V5.4.1.

My vs2022 version: Microsoft Visual Studio Community 2022 (64-bit) - Preview Version 17.3.0 Preview 1.1

I've tried lots of things but I'm not sure how much or relevant to describe here. Please let me know if I've missed anything that would be helpful. Anyone have any ideas on how to fix this "Java.Lang.NoClassDefFoundError"? 

[UdsMonitor.zip](https://github.com/dotnet/maui/files/8810829/UdsMonitor.zip)

**Comment by joshua211** [comment]
I have a similar problem where i get an `NoClassDefFoundError` exception for the class `Landroidx/camera/core/Preview$SurfaceProvider`.
Any news on this?

**Reply by losullivan59** [reply]
Hi Joshua, I did resolve my problem. The .aar library that I was trying to call from my maui app, was a few years old, and has some legacy android dependencies. I posted my question on stackoverflow, and someone recommended that my AAR had to use androidx rather than the deprecated android.support libraries. Since I had the source code to the AAR I was trying to call from maui, I was able to use android studio's automatic upgrading features to have it use androidx. After that, I was able to bring the aar into my maui app via visual studio's "Android Java Library Binding" project type. It sounds like you are already using androidx, though so your problem may be different?
